File: mod/orbit/dev/src/agent_base.py
"""
agent_base.py — class template for adding new agent backends to mod/dev.

Every agent module (claude, codex, cursor, …) should subclass `AgentBase`,
declare its CLI binary and a small set of provider-specific bits, and the
dev console will route to it automatically:

    from agent_base import AgentBase

    class Mod(AgentBase):
        NAME    = "cursor"
        ICON    = "▶"
        COLOR   = "#7c3aed"
        BINARY  = "cursor-agent"
        DEFAULT_MODEL = "claude-3.5-sonnet"
        ENV_KEY = "CURSOR_API_KEY"           # encrypted via wallet sig

        def build_args(self, prompt: str, model: str, work_dir: str) -> list[str]:
            return ["--print", "--model", model, prompt]

That's all that's strictly required. Optionally override `health()`,
`parse_stream()`, `bearer_token_format()`, etc. to handle CLI quirks.

The default `forward()`, `submit()`, `jobs()`, etc. all delegate to the
Rust job server (Cargo bin `<NAME>-jobs`) running on `self.port`. If your
agent doesn't have a Rust server yet, drop the binary in
`src/api/target/release/<name>-jobs` (or copy claude's and rebuild) — the
template doesn't care which binary, just that the convention is followed.
"""

import json
import os
import subprocess
import time
from pathlib import Path
from typing import Any, Dict, List, Optional

import mod as m


class AgentBase:
    # ── Override these in your subclass ──────────────────────────────
    NAME: str = ""               # module name, must match orbit dir
    ICON: str = ""               # 1-char or emoji shown in dev sidebar avatar
    COLOR: str = "#888888"       # hex color for sidebar avatar gradient
    BINARY: str = ""             # CLI binary the Rust server spawns
    DEFAULT_MODEL: str = ""      # model arg passed when caller omits one
    ENV_KEY: str = ""            # env var name for the provider's API key
    DESCRIPTION: str = "agent module (override DESCRIPTION in subclass)"

    # ── Optional overrides for CLI-specific behavior ─────────────────
    DANGEROUS_FLAG: Optional[str] = None  # e.g. "--dangerously-skip-permissions"

    def __init__(self):
        self._path = self._discover_path()
        self._cfg = self._load_config()

    # ── Subclass hook: build the argv for one prompt ─────────────────
    def build_args(self, prompt: str, model: str, work_dir: str) -> List[str]:
        """Return the CLI argv (excluding the binary itself) for a single
        prompt invocation. Default mirrors claude/codex shape; override if
        your CLI takes different flags.
        """
        args = []
        if self.DEFAULT_MODEL:
            args += ["--model", model or self.DEFAULT_MODEL]
        if self.DANGEROUS_FLAG:
            args.append(self.DANGEROUS_FLAG)
        args.append(prompt)
        return args

    # ── Defaults that work for any CLI agent — usually no override ───
    def _discover_path(self) -> str:
        here = Path(__file__).resolve().parent
        # Walk up looking for an orbit dir matching NAME
        for d in (here, *here.parents):
            p = d.parent / self.NAME
            if (p / "config.json").exists():
                return str(p)
        return str(here.parent.parent / self.NAME)

    def _load_config(self) -> dict:
        cp = Path(self._path) / "config.json"
        return json.loads(cp.read_text()) if cp.exists() else {}

    def forward(self, **kwargs):
        return self.info()

    def info(self) -> dict:
        return {
            "name": self.NAME,
            "icon": self.ICON,
            "color": self.COLOR,
            "binary": self.BINARY,
            "description": self.DESCRIPTION,
            "owner": self._cfg.get("owner"),
            "port": self._cfg.get("port"),
            "model": self.DEFAULT_MODEL,
            "env_key": self.ENV_KEY,
        }

    def health(self) -> dict:
        """Light liveness — overrides should hit the Rust API's /health."""
        port = self._cfg.get("port")
        if not port:
            return {"service": self.NAME, "status": "no port configured"}
        try:
            import urllib.request as r
            with r.urlopen(f"http://127.0.0.1:{port}/health", timeout=1) as resp:
                return json.loads(resp.read())
        except Exception as e:
            return {"service": self.NAME, "status": "down", "error": str(e)}

    def cli_path(self) -> Optional[str]:
        """Resolve the absolute path to the agent's CLI binary, or None."""
        if not self.BINARY:
            return None
        result = subprocess.run(["which", self.BINARY], capture_output=True, text=True)
        return result.stdout.strip() or None

    def submit(self, prompt: str, model: Optional[str] = None,
               work_dir: Optional[str] = None, key=None) -> dict:
        """Queue a job via the Rust job server, returning {job_id, …}.

        The Rust server reads provider info from this module's config.json
        and spawns BINARY with build_args(). Subclasses don't usually need
        to override this — just BINARY + build_args().
        """
        port = self._cfg.get("port")
        if not port:
            raise RuntimeError(f"{self.NAME}: no port in config.json")
        import urllib.request as r
        body = json.dumps({
            "prompt": prompt,
            "model": model or self.DEFAULT_MODEL,
            "work_dir": work_dir or os.getcwd(),
        }).encode()
        req = r.Request(f"http://127.0.0.1:{port}/jobs", method="POST", data=body,
                        headers={"Content-Type": "application/json"})
        if key:
            req.add_header("Authorization", f"Bearer {key}")
        try:
            with r.urlopen(req, timeout=30) as resp:
                return json.loads(resp.read())
        except Exception as e:
            return {"error": str(e)}

    def jobs(self) -> List[dict]:
        """List recent jobs from the Rust job server."""
        port = self._cfg.get("port")
        if not port:
            return []
        import urllib.request as r
        try:
            with r.urlopen(f"http://127.0.0.1:{port}/jobs", timeout=5) as resp:
                return json.loads(resp.read())
        except Exception:
            return []

    # ── Registry integration ─────────────────────────────────────────
    def cid(self) -> str:
        """Content CID for this module's tree, used by dev.merkle()."""
        try:
            return m.cid(self.NAME)
        except Exception:
            return ""

    def register(self, fund_eth: float = 0.0005, key=None) -> dict:
        """Register this agent on the mod protocol registry using its own
        deterministic key, with gas paid by the funded relayer.
        """
        dev = m.mod("dev")()
        return dev.register_per_key(modules=[self.NAME], fund_eth=fund_eth, key=key)

File: mod/orbit/dev/src/agent_base.js
// agent_base.js — JavaScript counterpart of agent_base.py
//
// Each agent module written in JS exports a class extending AgentContract,
// declares the same constants (NAME, BINARY, DEFAULT_MODEL, ENV_KEY), and
// optional view/tx methods. The same dispatcher in dev/src/api/main.py
// detects `contract.js` and shells out to `node contract.js <cmd> <json>`.

const fs = require("fs");
const path = require("path");
const crypto = require("crypto");
const { execSync, spawnSync } = require("child_process");

// Decorator-equivalent annotations applied via static metadata maps.
// In JS we declare a static `abi` map instead of decorators.
class AgentContract {
  // Subclasses override these (same names as the Python base):
  static NAME = "";
  static ICON = "";
  static COLOR = "#888888";
  static BINARY = "";
  static DEFAULT_MODEL = "";
  static ENV_KEY = "";
  static DESCRIPTION = "agent (override DESCRIPTION)";

  // Subclasses declare ABI as { name: { kind: "view"|"tx", ownerOnly: bool, doc: string } }
  static ABI = {
    info:    { kind: "view", ownerOnly: false, doc: "Return agent info" },
    health:  { kind: "view", ownerOnly: false, doc: "Probe agent backend" },
    submit:  { kind: "tx",   ownerOnly: false, doc: "Queue a prompt as a job" },
    jobs:    { kind: "view", ownerOnly: false, doc: "List recent jobs" },
  };

  constructor() {
    this._path = this._discoverPath();
    this._cfg = this._loadConfig();
    this._owner = (this._cfg.owner || "").toLowerCase();
    this.state = this._loadState();
  }

  static codeHash() {
    // Same scheme as Python AgentContract: sha3 over this file's bytes.
    const file = this._sourceFile || __filename;
    const buf = fs.readFileSync(file);
    return "0x" + crypto.createHash("sha3-256").update(buf).digest("hex");
  }

  static abi() {
    const out = [];
    for (const [name, meta] of Object.entries(this.ABI || {})) {
      out.push({ name, kind: meta.kind, owner_only: !!meta.ownerOnly, inputs: meta.inputs || [], doc: meta.doc });
    }
    return out.sort((a, b) => (a.kind + a.name).localeCompare(b.kind + b.name));
  }

  static manifest() {
    return {
      name: this.NAME,
      lang: "javascript",
      icon: this.ICON,
      color: this.COLOR,
      binary: this.BINARY,
      default_model: this.DEFAULT_MODEL,
      env_key: this.ENV_KEY,
      description: this.DESCRIPTION,
      code_hash: this.codeHash(),
      abi: this.abi(),
    };
  }

  _discoverPath() {
    const here = path.resolve(__dirname);
    let cur = here;
    for (let i = 0; i < 6; i++) {
      const cand = path.join(path.dirname(cur), this.constructor.NAME, "config.json");
      if (fs.existsSync(cand)) return path.dirname(cand);
      cur = path.dirname(cur);
    }
    return path.join(here, "..", "..", this.constructor.NAME);
  }

  _loadConfig() {
    const cp = path.join(this._path, "config.json");
    return fs.existsSync(cp) ? JSON.parse(fs.readFileSync(cp, "utf-8")) : {};
  }

  _statePath() {
    const dir = path.join(process.env.HOME, ".mod", this.constructor.NAME);
    fs.mkdirSync(dir, { recursive: true });
    return path.join(dir, "state.json");
  }

  _loadState() {
    const p = this._statePath();
    if (!fs.existsSync(p)) return { created_at: Math.floor(Date.now() / 1000), jobs_submitted: 0, events: 0 };
    try { return JSON.parse(fs.readFileSync(p, "utf-8")); } catch { return {}; }
  }

  _saveState() {
    fs.writeFileSync(this._statePath(), JSON.stringify(this.state, null, 2));
  }

  emit(name, fields = {}) {
    const evt = { event: name, ts: Math.floor(Date.now() / 1000), fields };
    this.state.events = (this.state.events || 0) + 1;
    const logPath = path.join(process.env.HOME, ".mod", this.constructor.NAME, "events.jsonl");
    fs.mkdirSync(path.dirname(logPath), { recursive: true });
    fs.appendFileSync(logPath, JSON.stringify(evt) + "\n");
  }

  info() {
    return {
      name: this.constructor.NAME,
      icon: this.constructor.ICON,
      color: this.constructor.COLOR,
      binary: this.constructor.BINARY,
      description: this.constructor.DESCRIPTION,
      owner: this._owner,
      port: this._cfg.port,
      default_model: this.constructor.DEFAULT_MODEL,
      env_key: this.constructor.ENV_KEY,
      code_hash: this.constructor.codeHash(),
      lang: "javascript",
    };
  }

  health() {
    const port = this._cfg.port;
    if (!port) return { service: this.constructor.NAME, status: "no port configured" };
    try {
      const out = execSync(`curl -sf http://127.0.0.1:${port}/health`, { timeout: 1000 }).toString();
      return JSON.parse(out);
    } catch (e) {
      return { service: this.constructor.NAME, status: "down" };
    }
  }

  cliPath() {
    if (!this.constructor.BINARY) return null;
    try {
      return execSync(`which ${this.constructor.BINARY}`).toString().trim() || null;
    } catch { return null; }
  }

  // Subclass hook — same role as Python build_args
  buildArgs(prompt, model, workDir) {
    const args = [];
    if (this.constructor.DEFAULT_MODEL) args.push("--model", model || this.constructor.DEFAULT_MODEL);
    args.push(prompt);
    return args;
  }

  jobs() {
    const port = this._cfg.port;
    if (!port) return [];
    try {
      const out = execSync(`curl -sf http://127.0.0.1:${port}/jobs`, { timeout: 5000 }).toString();
      return JSON.parse(out);
    } catch { return []; }
  }

  submit({ prompt, model, work_dir, key } = {}) {
    const port = this._cfg.port;
    if (!port) throw new Error(`${this.constructor.NAME}: no port`);
    const body = JSON.stringify({ prompt, model: model || this.constructor.DEFAULT_MODEL, work_dir: work_dir || process.cwd() });
    let res;
    try {
      const r = execSync(
        `curl -sf -X POST http://127.0.0.1:${port}/jobs -H 'Content-Type: application/json' ${key ? `-H 'Authorization: Bearer ${key}'` : ""} -d ${JSON.stringify(body)}`,
        { timeout: 30000 }
      ).toString();
      res = JSON.parse(r);
    } catch (e) {
      this.emit("submit_failed", { error: e.message });
      this._saveState();
      return { error: e.message };
    }
    this.state.jobs_submitted = (this.state.jobs_submitted || 0) + 1;
    this.emit("job_submitted", { job_id: res.id, prompt: (prompt || "").slice(0, 50), model });
    this._saveState();
    return res;
  }
}

// ── CLI dispatch — used by the Python dispatcher in dev/src/api/main.py
// Invoke as:  node contract.js manifest
//             node contract.js abi
//             node contract.js call <method> <jsonArgs>
function runCli(SubMod) {
  SubMod._sourceFile = require.main.filename;
  const [, , cmd, ...rest] = process.argv;
  try {
    if (cmd === "manifest") {
      console.log(JSON.stringify(SubMod.manifest()));
      return;
    }
    if (cmd === "abi") {
      console.log(JSON.stringify(SubMod.abi()));
      return;
    }
    if (cmd === "call") {
      const [methodName, argsJson] = rest;
      const args = argsJson ? JSON.parse(argsJson) : {};
      const inst = new SubMod();
      const fn = inst[methodName];
      if (typeof fn !== "function") throw new Error(`no method ${methodName}`);
      const result = fn.call(inst, args);
      console.log(JSON.stringify(result));
      return;
    }
    console.error(`usage: node ${path.basename(process.argv[1])} {manifest|abi|call <method> <jsonArgs>}`);
    process.exit(2);
  } catch (e) {
    console.error(JSON.stringify({ error: e.message }));
    process.exit(1);
  }
}

module.exports = { AgentContract, runCli };
